In [2]:
!pip install mlxtend

In [1]:
import numpy as np
import pandas as pd
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = 'all'

from mlxtend.frequent_patterns import apriori

from mlxtend.frequent_patterns import association_rules

In [5]:
df=pd.read_csv('C:/Users/G V Swetha/Desktop/PA.csv')

In [6]:
df


cardid    value pmethod sex homeown  income  age  fruitveg  freshmeat  \
0     39808  42.7123  CHEQUE   M      NO   27000   46         1          0   
1     67362  25.3567    CASH   F      NO   30000   28         1          0   
2     10872  20.6176    CASH   M      NO   13200   36         1          1   
3     26748  23.6883    CARD   F      NO   12200   26         1          1   
4     91609  18.8133    CARD   M     YES   11000   24         1          1   
..      ...      ...     ...  ..     ...     ...  ...       ...        ...   
995   31384  37.2056    CARD   M      NO   27500   27         1          1   
996   63996  13.7353  CHEQUE   F      NO   22600   23         1          1   
997   99025  29.0798    CARD   M     YES   27400   42         1          0   
998   95921  34.8576    CASH   F     YES   23300   43         0          1   
999   99164  30.6965    CASH   M      NO   21600   26         1          1   

     dairy  cannedveg  cannedmeat  frozenmeal  beer  wine  softdrink  fish  \
0        0          1           1           1     1     1          1     1   
1        1          1           1           1     1     1          1     1   
2        1          0           1           0     0     1          1     0   
3        0          1           1           1     1     0          1     1   
4        1          1           1           1     1     1          1     1   
..     ...        ...         ...         ...   ...   ...        ...   ...   
995      1          0           1           1     1     1          1     1   
996      1          0           1           1     1     1          1     0   
997      1          1           1           1     1     1          1     1   
998      1          1           1           1     1     0          1     1   
999      0          1           1           1     1     1          0     1   

     confectionery  
0                0  
1                0  
2                1  
3                1  
4                1  
..             ...  
995              1  
996              1  
997              1  
998              0  
999              0  

[1000 rows x 18 columns]

In [7]:
df.describe()

cardid        value        income          age     fruitveg  \
count    1000.000000  1000.000000   1000.000000  1000.000000  1000.000000   
mean    60405.704000    29.398240  20171.800000    32.840000     0.701000   
std     29494.202111    11.555366   5767.558365    10.342793     0.458049   
min     10150.000000    10.007000  10200.000000    16.000000     0.000000   
25%     34577.000000    19.057800  15275.000000    24.000000     0.000000   
50%     60629.500000    29.430550  20150.000000    32.000000     1.000000   
75%     87003.750000    39.127025  25100.000000    42.000000     1.000000   
max    109884.000000    49.886300  30000.000000    50.000000     1.000000   

        freshmeat       dairy    cannedveg   cannedmeat   frozenmeal  \
count  1000.00000  1000.00000  1000.000000  1000.000000  1000.000000   
mean      0.81700     0.82300     0.697000     0.796000     0.698000   
std       0.38686     0.38186     0.459785     0.403171     0.459355   
min       0.00000     0.00000     0.000000     0.000000     0.000000   
25%       1.00000     1.00000     0.000000     1.000000     0.000000   
50%       1.00000     1.00000     1.000000     1.000000     1.000000   
75%       1.00000     1.00000     1.000000     1.000000     1.000000   
max       1.00000     1.00000     1.000000     1.000000     1.000000   

              beer         wine    softdrink        fish  confectionery  
count  1000.000000  1000.000000  1000.000000  1000.00000     1000.00000  
mean      0.707000     0.713000     0.816000     0.70800        0.72400  
std       0.455366     0.452588     0.387678     0.45491        0.44724  
min       0.000000     0.000000     0.000000     0.00000        0.00000  
25%       0.000000     0.000000     1.000000     0.00000        0.00000  
50%       1.000000     1.000000     1.000000     1.00000        1.00000  
75%       1.000000     1.000000     1.000000     1.00000        1.00000  
max       1.000000     1.000000     1.000000     1.00000        1.00000

In [13]:
df.shape
df


(1000, 14)

pmethod sex homeown  fruitveg  freshmeat  dairy  cannedveg  cannedmeat  \
0    CHEQUE   M      NO         1          0      0          1           1   
1      CASH   F      NO         1          0      1          1           1   
2      CASH   M      NO         1          1      1          0           1   
3      CARD   F      NO         1          1      0          1           1   
4      CARD   M     YES         1          1      1          1           1   
..      ...  ..     ...       ...        ...    ...        ...         ...   
995    CARD   M      NO         1          1      1          0           1   
996  CHEQUE   F      NO         1          1      1          0           1   
997    CARD   M     YES         1          0      1          1           1   
998    CASH   F     YES         0          1      1          1           1   
999    CASH   M      NO         1          1      0          1           1   

     frozenmeal  beer  wine  softdrink  fish  confectionery  
0             1     1     1          1     1              0  
1             1     1     1          1     1              0  
2             0     0     1          1     0              1  
3             1     1     0          1     1              1  
4             1     1     1          1     1              1  
..          ...   ...   ...        ...   ...            ...  
995           1     1     1          1     1              1  
996           1     1     1          1     0              1  
997           1     1     1          1     1              1  
998           1     1     0          1     1              0  
999           1     1     1          0     1              0  

[1000 rows x 14 columns]

In [15]:
df.drop(['pmethod','homeown','sex'],axis=1,inplace=True)

In [11]:
df.shape
df

(1000, 14)

pmethod sex homeown  fruitveg  freshmeat  dairy  cannedveg  cannedmeat  \
0    CHEQUE   M      NO         1          0      0          1           1   
1      CASH   F      NO         1          0      1          1           1   
2      CASH   M      NO         1          1      1          0           1   
3      CARD   F      NO         1          1      0          1           1   
4      CARD   M     YES         1          1      1          1           1   
..      ...  ..     ...       ...        ...    ...        ...         ...   
995    CARD   M      NO         1          1      1          0           1   
996  CHEQUE   F      NO         1          1      1          0           1   
997    CARD   M     YES         1          0      1          1           1   
998    CASH   F     YES         0          1      1          1           1   
999    CASH   M      NO         1          1      0          1           1   

     frozenmeal  beer  wine  softdrink  fish  confectionery  
0             1     1     1          1     1              0  
1             1     1     1          1     1              0  
2             0     0     1          1     0              1  
3             1     1     0          1     1              1  
4             1     1     1          1     1              1  
..          ...   ...   ...        ...   ...            ...  
995           1     1     1          1     1              1  
996           1     1     1          1     0              1  
997           1     1     1          1     1              1  
998           1     1     0          1     1              0  
999           1     1     1          0     1              0  

[1000 rows x 14 columns]

In [16]:
df1=pd.get_dummies(df)
df1.head()

fruitveg  freshmeat  dairy  cannedveg  cannedmeat  frozenmeal  beer  wine  \
0         1          0      0          1           1           1     1     1   
1         1          0      1          1           1           1     1     1   
2         1          1      1          0           1           0     0     1   
3         1          1      0          1           1           1     1     0   
4         1          1      1          1           1           1     1     1   

   softdrink  fish  confectionery  
0          1     1              0  
1          1     1              0  
2          1     0              1  
3          1     1              1  
4          1     1              1

In [17]:
frequent_items=apriori(df1,min_support=0.05,use_colnames=True)
frequent_items.sort_values(by='support',ascending=False)

support                                           itemsets
2       0.823                                            (dairy)
1       0.817                                        (freshmeat)
8       0.816                                        (softdrink)
4       0.796                                       (cannedmeat)
10      0.724                                    (confectionery)
...       ...                                                ...
2041    0.074  (frozenmeal, fruitveg, softdrink, fish, confec...
2040    0.072  (fruitveg, softdrink, fish, confectionery, can...
2037    0.070  (frozenmeal, fruitveg, fish, confectionery, ca...
2035    0.069  (frozenmeal, fruitveg, softdrink, fish, canned...
2046    0.060  (frozenmeal, fruitveg, softdrink, fish, confec...

[2047 rows x 2 columns]

In [20]:
rules=association_rules(frequent_items,metric='lift',min_threshold=1)

In [21]:
type(rules)

pandas.core.frame.DataFrame

In [22]:
rules.sort_values(by='lift',ascending=True)

antecedents                      consequents  \
30                         (Product3_Egg)                 (Product2_Salsa)   
31                       (Product2_Salsa)                   (Product3_Egg)   
21                      (Product2_Butter)                 (Product3_Dairy)   
20                       (Product3_Dairy)                (Product2_Butter)   
24                         (Product2_Jam)                 (Product3_Dairy)   
..                                    ...                              ...   
48  (Product2_Vegetable, Product1_Fruits)                 (Product3_Salad)   
69                      (Product1_Nachos)   (Product3_Egg, Product2_Salsa)   
64         (Product3_Egg, Product2_Salsa)                (Product1_Nachos)   
65        (Product3_Egg, Product1_Nachos)                 (Product2_Salsa)   
68                       (Product2_Salsa)  (Product3_Egg, Product1_Nachos)   

    antecedent support  consequent support  support  confidence      lift  \
30              0.5760              0.1965   0.1175    0.203993  1.038133   
31              0.1965              0.5760   0.1175    0.597964  1.038133   
21              0.2465              0.2030   0.0520    0.210953  1.039179   
20              0.2030              0.2465   0.0520    0.256158  1.039179   
24              0.3410              0.2030   0.0720    0.211144  1.040117   
..                 ...                 ...      ...         ...       ...   
48              0.1705              0.2210   0.1705    1.000000  4.524887   
69              0.2200              0.1175   0.1175    0.534091  4.545455   
64              0.1175              0.2200   0.1175    1.000000  4.545455   
65              0.1175              0.1965   0.1175    1.000000  5.089059   
68              0.1965              0.1175   0.1175    0.597964  5.089059   

    leverage  conviction  
30  0.004316    1.009413  
31  0.004316    1.054633  
21  0.001960    1.010080  
20  0.001960    1.012983  
24  0.002777    1.010323  
..       ...         ...  
48  0.132820         inf  
69  0.091650    1.894146  
64  0.091650         inf  
65  0.094411         inf  
68  0.094411    2.195079  

[70 rows x 9 columns]

In [22]:
rules.describe()

antecedent support  consequent support        support     confidence  \
count       135610.000000       135610.000000  135610.000000  135610.000000   
mean             0.372620            0.372620       0.147004       0.427264   
std              0.148727            0.148727       0.060269       0.160528   
min              0.069000            0.069000       0.060000       0.073529   
25%              0.264000            0.264000       0.103000       0.304527   
50%              0.336000            0.336000       0.133000       0.404624   
75%              0.464000            0.464000       0.176000       0.530035   
max              0.823000            0.823000       0.675000       0.895652   

                lift       leverage     conviction  
count  135610.000000  135610.000000  135610.000000  
mean        1.158869       0.017712       1.118008  
std         0.112125       0.011621       0.123957  
min         1.000000       0.000000       1.000000  
25%         1.078725       0.008659       1.037299  
50%         1.143519       0.016593       1.081242  
75%         1.211311       0.024750       1.155464  
max         1.660697       0.098856       2.645000

In [28]:
rules[(rules['lift']>=1) & (rules['confidence']>=0.8) & (rules['support']>=0.2)]


antecedents      consequents  \
1                                             (fruitveg)      (freshmeat)   
3                                             (fruitveg)          (dairy)   
8                                             (fruitveg)      (softdrink)   
12                                           (freshmeat)          (dairy)   
13                                               (dairy)      (freshmeat)   
...                                                  ...              ...   
32116  (frozenmeal, confectionery, cannedveg, beer, w...      (softdrink)   
32424  (frozenmeal, softdrink, confectionery, cannedm...           (wine)   
32427    (frozenmeal, softdrink, cannedmeat, beer, wine)  (confectionery)   
32428  (frozenmeal, confectionery, cannedmeat, beer, ...      (softdrink)   
32429  (softdrink, confectionery, cannedmeat, beer, w...     (frozenmeal)   

       antecedent support  consequent support  support  confidence      lift  \
1                   0.701               0.817    0.577    0.823110  1.007478   
3                   0.701               0.823    0.586    0.835949  1.015733   
8                   0.701               0.816    0.573    0.817404  1.001720   
12                  0.817               0.823    0.673    0.823745  1.000906   
13                  0.823               0.817    0.673    0.817740  1.000906   
...                   ...                 ...      ...         ...       ...   
32116               0.271               0.816    0.224    0.826568  1.012951   
32424               0.263               0.713    0.211    0.802281  1.125219   
32427               0.263               0.724    0.211    0.802281  1.108123   
32428               0.253               0.816    0.211    0.833992  1.022049   
32429               0.260               0.698    0.211    0.811538  1.162663   

       leverage  conviction  
1      0.004283    1.034540  
3      0.009077    1.078930  
8      0.000984    1.007688  
12     0.000609    1.004229  
13     0.000609    1.004060  
...         ...         ...  
32116  0.002864    1.060936  
32424  0.023481    1.451558  
32427  0.020588    1.395923  
32428  0.004552    1.108381  
32429  0.029520    1.602449  

[2079 rows x 9 columns]

# Final Apriori rules considered are

Lift being > 1
Confidence is > 80%
Support > 20 %

Above list the display of products association which satisfy this rule :

1) There is 81% chance that people who purchased fruitveg  also purchased  freshmeat and lift is > 1 
2) Similarly there is 80% chance that folks who purchased (frozenmeal, softdrink, cannedmeat , beer, wine ) will purchase confectinary 